utility_analysis_new/dpengine.py

```diff
@@ -0,0 +1,145 @@
+# Copyright 2022 OpenMined.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""DPEngine and ContributionBounder for utility analysis."""
+import collections
+
+import pipeline_dp
+from pipeline_dp import DPEngine
+from pipeline_dp import DataExtractors
+from pipeline_dp import NoiseKind
+from pipeline_dp import contribution_bounders
+from pipeline_dp.contribution_bounders import ContributionBounder
+from dataclasses import dataclass
+
+
+class UtilityAnalysisDPEngine(DPEngine):
+
+    def __init__(self, budget_accountant: 'BudgetAccountant',
+                 backend: 'PipelineBackend'):
+        super().__init__(budget_accountant, backend)
+
+    def aggregate(self,
+                  col,
+                  params: pipeline_dp.AggregateParams,
+                  data_extractors: DataExtractors,
+                  public_partitions=None):
+        return super().aggregate(col, params, data_extractors,
+                                 public_partitions)
+
+    def _aggregate(self, col, params: pipeline_dp.AggregateParams,
+                   data_extractors: DataExtractors, public_partitions):
+        assert params.custom_combiners is None, "Cumtom combiners are not supported"
+        assert params.metrics == [
+            pipeline_dp.Metrics.COUNT
+        ], f"Supported only count metrics, metrics={params.metrics}"
+        assert public_partitions is not None, "Only public partitions supported"
+        assert params.contribution_bounds_already_enforced is False, "Utility Analysis when contribution bounds are already enforced is not supported"
+
+        # TODO: Implement custom combiner
```

Comment:
Custom combiners are combiners that can be implemented by PipelineDP users. Here it's probably meant utility analysis combiners.

Reply:
Yeah, that's what I meant to say. I'll be more clear